src/coreclr/jit/importer.cpp

```diff
@@ -2108,6 +2102,35 @@ GenTree* Compiler::impReadyToRunLookupToTree(CORINFO_CONST_LOOKUP* pLookup,
     return addr;
 }
 
+//------------------------------------------------------------------------
+// impIsProfileableCastHelper: Checks whether a tree is a cast helper suitable to
+//    to be profiled and then optimized with PGO data
+//
+// Arguments:
+//    tree - the tree object to check
+//
+// Returns:
+//    true if the tree is a cast helper of one of the suitable to profile types
+//
+bool Compiler::impIsProfileableCastHelper(GenTree* tree)
+{
+    if (!opts.OptimizationDisabled() || (JitConfig.JitCastProfiling() != 1))
```

Comment:
Avoid the double negative?
```suggestion
    if (opts.OptimizationEnabled() || (JitConfig.JitCastProfiling() != 1))
```
I wonder if this would be clearer as two separate methods, one to enable adding probes and another to enable checking if we might have added probes.

Then the "enable one" would return true if BBINSTR & (enable cast probes) and right kind of helper.

The "checking" one would return true if BBOPT & (enable cast probes) and right kind of helper.

Note OSR+PGO will instrument optimized code so might have both these return true in one compile. That might seem a bit odd but the OSR method extent can overlap the Tier0 extent, so there could common IL offsets between the two where both want to add a class probe. These should end up pointing at the same underlying class histogram.

Reply:
Also in the OSR case we won't probe if we expand inline...we should think about this case more carefully.

Reply:
Done. @AndyAyersMS Can you take another look?